components/taxonomy/src/values.rs

```diff
@@ -51,6 +55,189 @@ impl error::Error for TypeError {
+            Some(r) => Ok(r)
+        }
+    }
+
+    pub fn downcast<T>(&self) -> Option<&T> where T: Data {
+        self.content.data.downcast_ref::<T>()
+    }
+
+    pub fn description(&self) -> String {
+        (self.content.describe)()
+    }
+}
+
+impl PartialEq for Value {
+    fn eq(&self, other: &Value) -> bool {
+        (self.content.eq)(&*self.content.data, &*other.content.data)
+    }
+}
+
+pub trait Data: Debug + Send + Sync + mopa::Any {
+    /// A human-readable description of the _type_ of the value.
+    ///
+    /// Used mainly in `TypeError` error messages.
+    fn description() -> String where Self: Sized;
+
+    /// Attempt to build a `Value` from a json `source` and `binary` components.
+    fn parse(path: Path, source: &JSON, binary: &BinarySource) -> Result<Self, Error> where Self: Sized;
+
+    /// Serialize a `Value` into a `JSON`, storing binary data in `binary`.
+    fn serialize(source: &Self, binary: &BinaryTarget) -> Result<JSON, Error> where Self: Sized;
+
+    /// Shorthand for parsing from a string.
+    ///
+    /// Used mainly for testing purposes.
+    fn parse_str(source: &str) -> Result<Self, Error> where Self: Sized {
+        serde_json::from_str(source)
+            .map_err(|err| Error::ParseError(ParseError::JSON(JSONError(err))))
+            .and_then(|json| Self::parse(Path::new(), &json, &BinarySource))
+    }
+
+    fn parse_vec(path: Path, source: &JSON, binary: &BinarySource) -> Result<Vec<Self>, Error> where Self: Sized {
+        match source.as_array() {
+            None => Err(Error::TypeError(TypeError {
+                expected: "array".to_owned(),
+                got: "something else".to_owned()
+            })),
+            Some(array) => {
+                let mut result = Vec::with_capacity(array.len());
+                for (item, i) in array.iter().zip(0..) {
+                    let got = try!(path.push_index(i, |path| Self::parse(path, item, binary)));
+                    result.push(got);
+                }
+                Ok(result)
+            }
+        }
+    }
+
+    fn parse_field(path: Path, source: &JSON, binary: &BinarySource, field_name: &str) -> Result<Self, Error> where Self: Sized {
+        match Self::parse_opt_field(path.clone(), source, binary, field_name) { // FIXME: Get rid of this `path.clone()`
+            Some(result) => result,
+            None => Err(Error::ParseError(ParseError::missing_field(field_name, &path)))
+        }
+    }
+
+    fn parse_opt_field(path: Path, source: &JSON, binary: &BinarySource, field_name: &str) -> Option<Result<Self, Error>> where Self: Sized {
+        if let JSON::Object(ref obj) = *source {
+            if let Some(v) = obj.get(field_name) {
+                Some(Self::parse(path, v, binary))
+            } else {
+                None
+            }
+        } else {
+            Some(Err(Error::ParseError(ParseError::type_error(field_name, &path, "object"))))
+        }
+    }
+}
+mopafy!(Data);
```

Comment:
so that I understand properly, this is how we'll let adapters implement their own data types ?
Mopa is like an "open" enum ?